### PR TITLE
lib: support security prepare

### PR DESCRIPTION
**Pull request by RafaelGSS**
I'm doing an experiment as part of my work on https://github.com/nodejs/security-wg/issues/860 + upcoming security release.

I was talking with @BethGriggs and looks reasonable to generate the proposal PR for security releases using `git node release --security --prepare 19.6.1`.

Currently, `git node land` doesn't work for the private repo, so the workflow is:

1. Create the proposal branch (based on vX.X instead of staging)
2. Iterate open PRs (based on the given filter) and ask to amend them in the proposal
3. Proceed with the regular workflow (adjust the node_version, changelogs, ...)
 
Example:

```console
➜  node-private (main) ../node-core-utils/bin/git-node.js release --prepare --security 20.2.1
   ⚠  Remote repository URL does not point to the expected repository nodejs/node
   ⚠  You are trying to create a new release proposal branch for v20, but you're checked out on main and not v20.x.
--------------------------------------------------------------------------------
   ℹ  Switch to `v20.x` with `git checkout v20.x` before proceeding.
--------------------------------------------------------------------------------
➜  node-private (main) git checkout v20.x 
Switched to branch 'v20.x'
Your branch is up to date with 'upstream/v20.x'.
➜  node-private (v20.x) ../node-core-utils/bin/git-node.js release --prepare --security 20.2.1
   ⚠  Remote repository URL does not point to the expected repository nodejs/node
⠙ Creating new proposal branch for 20.2.1Branch 'v20.2.1-proposal' set up to track remote branch 'v20.x' from 'origin'.
Switched to a new branch 'v20.2.1-proposal'
✔  Created new proposal branch for 20.2.1
✔  Done loading data for nodejs-private/node-private/pull/393
----------------------------------- PR info ------------------------------------
Title      SENSITIVE: INFORMATION (#393)
   ⚠  Could not retrieve the email or name of the PR author's from user's GitHub profile!
Branch     tniessen:h1-XXXXX -> nodejs-private:main
Labels
Commits    1
 - sensitive: information
Committers 1
 - Tobias Nießen <tniessen@tnie.de>
------------------------------ Generated metadata ------------------------------
PR-URL: https://github.com/nodejs-private/node-private/pull/393
Fixes: X
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
--------------------------------------------------------------------------------
   ℹ  This PR was created on Thu, 23 Feb 2023 15:18:32 GMT
   ✔  Approvals: 3
   ✔  - Matteo Collina (@mcollina) (TSC): https://github.com/nodejs-private/node-private/pull/393#pullrequestreview-1352725924
   ✔  - Robert Nagy (@ronag) (TSC): https://github.com/nodejs-private/node-private/pull/393#pullrequestreview-1311573538
   ✔  - Rafael Gonzaga (@RafaelGSS) (TSC): https://github.com/nodejs-private/node-private/pull/393#pullrequestreview-1311635021
   ✖  Last GitHub CI failed
   ✖  No Jenkins CI runs detected
--------------------------------------------------------------------------------
? Would you like to amend this PR to the proposal? Yes
⠙ Downloading patch for 393From github.com:nodejs-private/node-private
 * branch                    refs/pull/393/merge -> FETCH_HEAD
✔  Fetched commits as 745083abb4a7..88b780e6e074
--------------------------------------------------------------------------------
[v20.2.1-proposal d79ce3ef3a4] sensitive: information
 Author: Tobias Nießen <tniessen@tnie.de>
 Date: Thu Feb 23 15:13:16 2023 +0000
 2 files changed, 43 insertions(+)
   ✔  Patches applied
--------------------------------------------------------------------------------
? Git found no CVE-ID trailer in the original commit message. Please, provide the CVE-ID CVE-2023-TEST123
--------------------------------- New Message ----------------------------------
sesntive: information

REMOVED DESCRIPTION.

Fixes: X
PR-URL: https://github.com/nodejs-private/node-private/pull/393
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
CVE-ID: CVE-2023-TEST123
--------------------------------------------------------------------------------
? Use this message? Yes
```

---

Limitations that should be addressed in subsequent PRs:

- Support multiple commits
- Public PRs (backports)

### CODE REVIEW

**Comment by RafaelGSS** [comment]
@nodejs/releasers just updated this PR. 

Now it iterates over open PRs in the private repo and adds them to the proposal branch. If the PR doesn't contain CVE-ID metadata, it will ask to include it in the middle of the process. Could you please review it?

File: lib/prepare_release.js [hunk 1]
```diff
@@ -67,9 +72,144 @@ export default class ReleasePreparation {
...
+    cli.startSpinner('Updating CHANGELOG.md');
+    await this.updateMainChangelog();
+    cli.stopSpinner('Updated CHANGELOG.md');
+
+    cli.startSpinner(`Updating CHANGELOG_V${versionComponents.major}.md`);
+    await this.updateMajorChangelog();
+    cli.stopSpinner(`Updated CHANGELOG_V${versionComponents.major}.md`);
+
+    await cli.prompt('Finished editing the changelogs?',
+      { defaultAnswer: false });
```
Comment (ruyadorno) [inline lib/prepare_release.js:167]: not sure what this prompt is supposed to do?

File: lib/prepare_release.js [hunk 2]
```diff
@@ -67,9 +72,144 @@ export default class ReleasePreparation {
...
+    // it to their liking.
+    const createDefaultCommit = await this.createReleaseCommit();
+    if (!createDefaultCommit) {
+      const lastCommitSha = runSync('git', ['rev-parse', '--short', 'HEAD']);
+      cli.warn(`Please manually edit commit ${lastCommitSha} by running ` +
+      '`git commit --amend` before proceeding.');
+
+      await cli.prompt(
+        'Finished editing the release commit?',
+        { defaultAnswer: false });
```
Comment (ruyadorno) [inline lib/prepare_release.js:187]: does `cli.prompt` throws if the answer is negative or something? I'm not sure what this prompt here is doing either 🤔

File: lib/prepare_release.js [hunk 3]
```diff
@@ -382,8 +523,11 @@ export default class ReleasePreparation {
     const arr = data.split('\n');
     const allCommits = this.getChangelog();
     const notableChanges = this.getBranchDiff({ onlyNotableChanges: true });
-    const releaseHeader = `## ${date}, Version ${newVersion}` +
+    let releaseHeader = `## ${date}, Version ${newVersion}` +
                           ` ${releaseInfo}, @${username}\n`;
+    if (isSecurityRelease) {
+      releaseHeader += 'This is a security release.';
```
Comment (ruyadorno) [inline lib/prepare_release.js:529]: I think it might need an extra line break:

```suggestion
      releaseHeader += '\nThis is a security release.';
```

**Review by ruyadorno — COMMENTED** [review]
LGTM! A few details you might want to clean up before shipping (although I might just need clarification in a few places) 👍

Great work! Security releases are going to be much simpler 😊 🙏

File: lib/cherry_pick.js [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    const subjects = await runAsync('git',
+      ['log', '--pretty=format:%s', `${commitInfo.base}..${commitInfo.head}`],
+      { captureStdout: 'lines' });
+
+    if (commitInfo.shas.length !== 1) {
+      const fixupAll = await cli.prompt(
+        `There are ${subjects.length} commits in the PR. ` +
+        'Would you like to fixup everything into first commit?');
+      if (!fixupAll) {
+        // TODO: add this support?
```
[the remark this answers is not in the document]
Comment (RafaelGSS) [reply lib/cherry_pick.js:196]: Since we don't have sessions, I think that's the only approach...

[the remark this answers is not in the document]
Comment (ruyadorno) [reply lib/cherry_pick.js:196]: I'm def fine with that 😊

File: lib/prepare_release.js (hunk 2 above)
Comment (RafaelGSS) [reply lib/prepare_release.js:187]: it was just a copy-paste from the regular release process :). Fixed, thanks

**Review by ruyadorno — APPROVED** [review]

**Comment by ruyadorno** [comment]
🎉

**Pull request merged**